### PR TITLE
[csi-plugins] add support for PVC annotations

### LINKED ISSUE #2685: [manila-csi-plugin] add support for share group and affinity annotations
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature

**What happened**:

The Manila API allows to specify a share group in which the new share should be created. Along with this it supports affinity/anti-affinity scheduler hints.

**What you expected to happen**:

The share group ID can be specified in manila storage class:

```yaml
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: csi-manila-nfs
provisioner: nfs.manila.csi.openstack.org
allowVolumeExpansion: true
parameters:
  # Manila share type
  type: default
  # Manila share group
  group-id: %my-group-id%
```

The scheduler hints for a particular share can be specified in a PVC annotation like it's donw in AWS EBS CSI driver:

```
$ k describe pv           
Name:              pvc-6bff47f9-1843-4576-ba15-158c73491e8c
Labels:            <none>
Annotations:       ebs.csi.aws.com/iops: 4000
                   ebs.csi.aws.com/volumeType: io2
```

The proposed annotation keys are:

* `manila.csi.openstack.org/affinity: %shareID1%,%shareID2%`
* `manila.csi.openstack.org/anti-affinity: %shareID3%,%shareID4%`
* `manila.csi.openstack.org/group-id: %groupID%` - will override the SC group ID

The PVC with the annotations below:

```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  annotations:
    manila.csi.openstack.org/affinity: a07a01c7-43b9-484a-8745-717e44d8f662,749ad593-0a99-4e32-91ba-7106db7ef792
    manila.csi.openstack.org/anti-affinity: 364f7c8c-7a60-4023-8458-b332e176789a,1101d1d5-b68d-4f17-8604-46271852b6db
    manila.csi.openstack.org/group-id: 5b5f4dce-6437-461e-80bb-77b0a1f3671f
```

and storage class:

```
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: csi-manila-nfs
provisioner: nfs.manila.csi.openstack.org
allowVolumeExpansion: true
parameters:
  type: default
  group-id: ca8e84aa-a0c7-4631-9b82-69c95b50a4bd
```

will produce an API call below:

```
{
    "share": {
        "share_proto": "nfs",
        "share_network_id": "713df749-aac0-4a54-af52-10f6c991e80c",
        "share_group_id": "5b5f4dce-6437-461e-80bb-77b0a1f3671f",
        "name": "pvc_123",
        "size": 1,
        "scheduler_hints": {
            "same_host": "a07a01c7-43b9-484a-8745-717e44d8f662,749ad593-0a99-4e32-91ba-7106db7ef792",
            "different_host": "364f7c8c-7a60-4023-8458-b332e176789a,1101d1d5-b68d-4f17-8604-46271852b6db"
        }
    }
}
```

Accessing PVC annotations can be supported with a side-car modifier, see https://github.com/kubernetes-sigs/aws-ebs-csi-driver/pull/1600 for example.

**How to reproduce it**:

n/a

**Anything else we need to know?**:

Upstream gophercloud feature request for the scheduler hints: https://github.com/gophercloud/gophercloud/issues/3204

Additionally we can implement the [VolumeAttributesClass](https://kubernetes.io/docs/concepts/storage/volume-attributes-classes/) support. `VolumeAttributesClass`  parameters will override `StorageClass` parameters, and PVC annotations will override `VolumeAttributesClass` parameters. For the `volumeAttributesClassName` support manila CSI must have a k8s client configured. However this can be an extra task. See https://github.com/kubernetes-sigs/aws-ebs-csi-driver/pull/1941 for reference. This approach won't require a side-car container to fetch PVC annotations.

**Environment**:
- openstack-cloud-controller-manager(or other related binary) version:
- OpenStack version:
- Others:

### PR DESCRIPTION
**Pull request by kayrus**
**What this PR does / why we need it**:

This PR adds an ability to use PVC annotations to create persistent volumes with extra options. 

For manila-csi-plugin a new `groupID` option is added allowing the share to be created in a specific group.

**Which issue this PR fixes(if applicable)**:
fixes #2685 

**Special notes for reviewers**:

This is a fraft, since it's depends on:
* https://github.com/gophercloud/gophercloud/pull/3211
* https://github.com/gophercloud/gophercloud/pull/3205

**Release note**:

```release-note
[csi-plugins] add support for PVC annotations
[manila-csi-plugin] add support for share group id
```

### CODE REVIEW

**Comment by kayrus** [comment]
@zetaab @dulek @jichenjc ready for review

**Comment by kayrus** [comment]
I just discovered that `var  *volumes.SchedulerHintOpts` is not nil, when the func receives the interface as an arg.

```
if schedulerHints != nil {
   schedulerHints.ToSchedulerHintsMap()
}
```

> this will cause a `panic: value method openstack/blockstorage/v3/volumes.SchedulerHintOpts.ToSchedulerHintsMap called using nil *SchedulerHintOpts pointer`

According to go, the passed `var  *volumes.SchedulerHintOpts` is `(*volumes.SchedulerHintOpts)(nil)`, and passed `var volumes.SchedulerHintOptsBuilder` interface is `<nil>`

mind-blown-mind

So I had to push this:

```diff
diff --git a/pkg/csi/cinder/controllerserver.go b/pkg/csi/cinder/controllerserver.go
index a96b817a..dd630ba8 100644
--- a/pkg/csi/cinder/controllerserver.go
+++ b/pkg/csi/cinder/controllerserver.go
@@ -198,7 +198,7 @@ func (cs *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
        }
 
        // Set scheduler hints if affinity or anti-affinity is set in PVC annotations
-       var schedulerHints *volumes.SchedulerHintOpts
+       var schedulerHints volumes.SchedulerHintOptsBuilder
        var volCtx map[string]string
        affinity := util.SplitTrimJoin(pvcAnnotations[affinityKey], ',')
        antiAffinity := util.SplitTrimJoin(pvcAnnotations[antiAffinityKey], ',')
diff --git a/pkg/csi/cinder/openstack/openstack.go b/pkg/csi/cinder/openstack/openstack.go
index cc733690..46116ecb 100644
--- a/pkg/csi/cinder/openstack/openstack.go
+++ b/pkg/csi/cinder/openstack/openstack.go
@@ -45,7 +45,7 @@ func AddExtraFlags(fs *pflag.FlagSet) {
 }
 
 type IOpenStack interface {
-       CreateVolume(*volumes.CreateOpts, *volumes.SchedulerHintOpts) (*volumes.Volume, error)
+       CreateVolume(*volumes.CreateOpts, volumes.SchedulerHintOptsBuilder) (*volumes.Volume, error)
        DeleteVolume(volumeID string) error
        AttachVolume(instanceID, volumeID string) (string, error)
        ListVolumes(limit int, startingToken string) ([]volumes.Volume, string, error)
diff --git a/pkg/csi/cinder/openstack/openstack_mock.go b/pkg/csi/cinder/openstack/openstack_mock.go
index b35d923c..89f0d3dc 100644
--- a/pkg/csi/cinder/openstack/openstack_mock.go
+++ b/pkg/csi/cinder/openstack/openstack_mock.go
@@ -85,7 +85,7 @@ func (_m *OpenStackMock) AttachVolume(instanceID string, volumeID string) (strin
 }
 
 // CreateVolume provides a mock function with given fields: name, size, vtype, availability, tags
-func (_m *OpenStackMock) CreateVolume(opts *volumes.CreateOpts, _ *volumes.SchedulerHintOpts) (*volumes.Volume, error) {
+func (_m *OpenStackMock) CreateVolume(opts *volumes.CreateOpts, _ volumes.SchedulerHintOptsBuilder) (*volumes.Volume, error) {
        name := opts.Name
        size := opts.Size
        vtype := opts.VolumeType
diff --git a/pkg/csi/cinder/openstack/openstack_volumes.go b/pkg/csi/cinder/openstack/openstack_volumes.go
index f7f6be95..b3d90230 100644
--- a/pkg/csi/cinder/openstack/openstack_volumes.go
+++ b/pkg/csi/cinder/openstack/openstack_volumes.go
@@ -51,7 +51,7 @@ const (
 var volumeErrorStates = [...]string{"error", "error_extending", "error_deleting"}
 
 // CreateVolume creates a volume of given size
-func (os *OpenStack) CreateVolume(opts *volumes.CreateOpts, schedulerHints *volumes.SchedulerHintOpts) (*volumes.Volume, error) {
+func (os *OpenStack) CreateVolume(opts *volumes.CreateOpts, schedulerHints volumes.SchedulerHintOptsBuilder) (*volumes.Volume, error) {
        blockstorageClient, err := openstack.NewBlockStorageV3(os.blockstorage.ProviderClient, os.epOpts)
        if err != nil {
                return nil, err
diff --git a/tests/sanity/cinder/fakecloud.go b/tests/sanity/cinder/fakecloud.go
index a02a8e77..56ee04ad 100644
--- a/tests/sanity/cinder/fakecloud.go
+++ b/tests/sanity/cinder/fakecloud.go
@@ -34,7 +34,7 @@ func getfakecloud() *cloud {
 var _ openstack.IOpenStack = &cloud{}
 
 // Fake Cloud
-func (cloud *cloud) CreateVolume(opts *volumes.CreateOpts, _ *volumes.SchedulerHintOpts) (*volumes.Volume, error) {
+func (cloud *cloud) CreateVolume(opts *volumes.CreateOpts, _ volumes.SchedulerHintOptsBuilder) (*volumes.Volume, error) {
        vol := &volumes.Volume{
                ID:               randString(10),
                Name:             opts.Name,
```

to fix the:

```
panic: value method github.com/gophercloud/gophercloud/v2/openstack/blockstorage/v3/volumes.SchedulerHintOpts.ToSchedulerHintsMap called using nil *SchedulerHintOpts pointer

goroutine 62 [running]:
github.com/gophercloud/gophercloud/v2/openstack/blockstorage/v3/volumes.(*SchedulerHintOpts).ToSchedulerHintsMap(0xd?)
	<autogenerated>:1 +0x8c
github.com/gophercloud/gophercloud/v2/openstack/blockstorage/v3/volumes.Create({0x22647c8, 0x3383fc0}, 0xc000233180, {0x2241820, 0xc00002e370}, {0x2241840, 0x0})
	github.com/gophercloud/gophercloud/v2@v2.1.2-0.20241016132526-1c4dd03733fe/openstack/blockstorage/v3/volumes/requests.go:154 +0xf7
k8s.io/cloud-provider-openstack/pkg/csi/cinder/openstack.(*OpenStack).CreateVolume(0xc0001a6c80, 0xc00002e370, 0x0)
	k8s.io/cloud-provider-openstack/pkg/csi/cinder/openstack/openstack_volumes.go:68 +0x245
k8s.io/cloud-provider-openstack/pkg/csi/cinder.(*controllerServer).CreateVolume(0xc00051a030, {0x1ec1a20?, 0x20?}, 0xc0000c6800)
	k8s.io/cloud-provider-openstack/pkg/csi/cinder/controllerserver.go:215 +0x1caf
github.com/container-storage-interface/spec/lib/go/csi._Controller_CreateVolume_Handler.func1({0x2264800?, 0xc0003c3500?}, {0x1e83680?, 0xc0000c6800?})
	github.com/container-storage-interface/spec@v1.9.0/lib/go/csi/csi.pb.go:6587 +0xcb
k8s.io/cloud-provider-openstack/pkg/csi/cinder.logGRPC({0x2264800, 0xc0003c3500}, {0x1e83680, 0xc0000c6800}, 0xc000307240, 0xc000415c38)
	k8s.io/cloud-provider-openstack/pkg/csi/cinder/utils.go:97 +0x2f5
github.com/container-storage-interface/spec/lib/go/csi._Controller_CreateVolume_Handler({0x1ec1a20, 0xc00051a030}, {0x2264800, 0xc0003c3500}, 0xc0000c6700, 0x2089b60)
	github.com/container-storage-interface/spec@v1.9.0/lib/go/csi/csi.pb.go:6589 +0x143
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0001a3000, {0x2264800, 0xc0003c3470}, {0x2270ae0, 0xc0004e9680}, 0xc00017b200, 0xc000502b40, 0x3306020, 0x0)
	google.golang.org/grpc@v1.65.0/server.go:1379 +0xdf8
google.golang.org/grpc.(*Server).handleStream(0xc0001a3000, {0x2270ae0, 0xc0004e9680}, 0xc00017b200)
	google.golang.org/grpc@v1.65.0/server.go:1790 +0xe8b
google.golang.org/grpc.(*Server).serveStreams.func2.1()
	google.golang.org/grpc@v1.65.0/server.go:1029 +0x8b
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 60
	google.golang.org/grpc@v1.65.0/server.go:104
```

**Comment by kayrus** [comment]
@zetaab @dulek @jichenjc kindly ping

**Comment by kayrus** [comment]
@zetaab @dulek @jichenjc do you have any updates?

File: docs/cinder-csi-plugin/using-cinder-csi-plugin.md [hunk 1]
```diff
@@ -110,6 +111,14 @@ In addition to the standard set of klog flags, `cinder-csi-plugin` accepts the f
   If set to true then the CSI driver does provide the node service.
 
   The default is to provide the node service.
+
+  <dt>--pvc-annotations &lt;disabled&gt;</dt>
+  <dd>
+  If set to true then the CSI driver will use PVC annotations to provide volume
+  sceduler hints. See [Supported PVC Annotations](#supported-pvc-annotations)
```
Comment (dulek) [inline docs/cinder-csi-plugin/using-cinder-csi-plugin.md:118]: scheduler

**Review by dulek — COMMENTED** [review]
Do I correctly understand that this requires knowing the volume or share ID upfront before creating a PVC? This is totally not a K8s pattern, as to e.g. deploy three volumes with anti-affinity you need to create one, wait for ID to be populated, create second, wait for its ID and only then create the third, right?

I can see that for Manila you can specify share names too? When looking at the filters code I don't exactly see that support, but nevertheless:

> With scheduler hints, you can optionally specify the affinity and anti-affinity rules in relation to other shares. The scheduler will enforce these rules when determining where to create the share. Possible keys are same_host and different_host, and the value must be the share name or id.

Source: https://docs.openstack.org/manila/latest/user/create-and-manage-shares.html#create-a-share

Does this work? Name's easier to use, so I think it would be great to implement and document support for that (by implement I mean - do we need to do some names manipulations to map K8s objects to Manila shares? If so users should specify annotations using K8s names).

**Comment by kayrus** [comment]
@dulek 

> Do I correctly understand that this requires knowing the volume or share ID upfront before creating a PVC? This is totally not a K8s pattern, as to e.g. deploy three volumes with anti-affinity you need to create one, wait for ID to be populated, create second, wait for its ID and only then create the third, right?

Right. Unfortunately I don't see any other way to distribute shares across multiple backends. There will be a [feature](https://review.opendev.org/c/openstack/manila-specs/+/933558), based on the share group id, but I suppose it will be implemented mid next year, that's why I introduced a group ID support in advance.

> the value must be the share name or id

Interesting finding, but it doesn't work with API. I tested this locally and got: `{"badRequest": {"code": 400, "message": "123a is not a valid uuid."}}`. Do you think it's worth adding the resolving share ID by name like it's done in user CLI tool?

**Comment by dulek** [comment]
Alright, so we get what we can here.

/lgtm
/approve
/hold for @kayrus to decide when this is ready.

**Comment by kayrus** [comment]
@dulek waiting for a second lgtm

**Pull request merged**